### ISSUE TITLE
Suggestion: Handle incoming calls asynchronously

### ISSUE THREAD

**Issue by markvandenbergh** [ocpi/ocpi#198]
Suggestion for OCPI 2.2:

It would be great if we could process calls asynchronously. Because of the response code we first have to check if we have the object :
```
When the server receives a valid OCPI object it should respond with:

HTTP 200 - Ok when the object already existed and is successfully updated.
HTTP 201 - Created when the object is newly created in the server system.
```

If we could respond with "I understand your message, all looks fine and I'm going to process this asap" (without letting the sender know if I already have the object) that would make it easier to handle a higher load.

If we end up not having the object we can schedule a GET to retrieve it at the sender.

**Comment by poohsen** [comment]
two things: 
1) responding with the right HTTP code in such a case is more of a soft requirement. I wonder if anybody's implementation actually depends on the server making this difference.

2) That said, the only part of OCPI where I expect the rate of change to become a challenge performance-wise is the EVSE status and the meter value in a session. In general the bulk of data should be pulled, not pushed. The PUSH is there to enable near-realtime availability of data.

It's still possible that some implementors e.g. PUSH a whole EVSE object instead of just changing the status via a PATCH. However, I have proposed a change that would also affect the way the status and meter values are changed: #146. 

And like I said, for the rest of the domain objects I would claim that performance shouldn't play a big role. In which module are you experiencing the performance issues?

**Comment by markvandenbergh** [comment]
> responding with the right HTTP code in such a case is more of a soft requirement. I wonder if anybody's implementation actually depends on the server making this difference.

If the specification says there is a difference the best would be to follow it I assume. If nobody is using the difference it would simplify implementation to specify "HTTP 200" as the response to give in all "ok" situations.

> In general the bulk of data should be pulled, not pushed.

Not sure if I agree with you there. I think of the GET as a fail-safe. Just to retrieve the info we might have missed during downtime or network issues.

> That said, the only part of OCPI where I expect the rate of change to become a challenge performance-wise is the EVSE status and the meter value in a session.

Yes I agree.

> In which module are you experiencing the performance issues?

I'm currently not experiencing performance issues.

In my opinion it would simplify implementation if I can handle the incoming call by immediately responding with HTTP 200 without the need to process it before responding.

That would be possible if the spec would remove the part of "200" vs "201" as you mention is not really set in stone.. ?

> responding with the right HTTP code in such a case is more of a soft requirement.

**Comment by poohsen** [comment]
> If the specification says there is a difference the best would be to follow it I assume. If nobody is using the difference it would simplify implementation to specify "HTTP 200" as the response to give in all "ok" situations.

you're right. I looked it up and it's actually quite explicit. Although from our own implementation experience I can say that we do not act on the 200/201 difference in any way as a client.

>> In general the bulk of data should be pulled, not pushed.
> I think of the GET as a fail-safe. Just to retrieve the info we might have missed during downtime or network issues.

That, and also to get the complete set of data when you connect to a module for the first time. 
Maybe my formulation was a little misleading: of course once you correctly connected  you should receive all updates as a PUSH (if supported by the partner). But the pushes will be spread out over a longer time so they should not cause performance issues. There shouldn't be situations where you receive thousands of requests in bulk.

I would be interested in hearing other opinions on the usefulness of the 200/201 distinction anyway.

**Comment by RobertdeLeeuw-IHomer** [comment]
During the Strasbourg meeting this was discussed, current proposal for 3.0 is to replace GET All with an async GET/Push:
Client sends a GET
Server accepts the request.
Server sends batches of Objects via Push (BTW Batch Push is also an idea for OCPI 3.0)

This allows the server to handle large amounts of data more convenient.
